src/envoy/auth/config.cc

```diff
@@ -13,200 +13,97 @@
+  issuer->name = json.getString("name", "");
+  if (issuer->name == "") {
+    throw EnvoyException("Issuer name missing");
   }
+
+  // Check "audience".
+  // It will be an empty array if the key "audience" does not exist.
+  auto audiences = json.getStringArray("audiences", true);
+  issuer->audiences.insert(audiences.begin(), audiences.end());
+
   // Check "pubkey"
-  Json::ObjectSharedPtr json_pubkey;
-  try {
-    json_pubkey = json->getObject("pubkey");
-  } catch (...) {
-    ENVOY_LOG(debug, "IssuerInfo [name = {}]: Public key missing", name_);
-    failed_ = true;
-    return;
-  }
+  auto json_pubkey = json.getObject("pubkey");
+
   // Check "type"
-  std::string pkey_type_str = json_pubkey->getString("type", "");
-  if (pkey_type_str == "pem") {
-    pkey_type_ = Pubkeys::PEM;
-  } else if (pkey_type_str == "jwks") {
-    pkey_type_ = Pubkeys::JWKS;
+  std::string pubkey_type_str = json_pubkey->getString("type", "");
+  if (pubkey_type_str == "pem") {
+    issuer->pubkey_type = Pubkeys::PEM;
+  } else if (pubkey_type_str == "jwks") {
+    issuer->pubkey_type = Pubkeys::JWKS;
   } else {
-    ENVOY_LOG(debug,
-              "IssuerInfo [name = {}]: Public key type missing or invalid",
-              name_);
-    failed_ = true;
-    return;
+    throw EnvoyException(
+        fmt::format("Issuer [name = {}]: Public key type missing or invalid",
+                    issuer->name));
   }
+
   // Check "value"
-  std::string value = json_pubkey->getString("value", "");
-  if (value != "") {
-    // Public key is written in this JSON.
-    pkey_ = Pubkeys::CreateFrom(value, pkey_type_);
-    loaded_ = true;
-    return;
-  }
-  // Check "file"
-  std::string path = json_pubkey->getString("file", "");
-  if (path != "") {
-    // Public key is loaded from the specified file.
-    pkey_ = Pubkeys::CreateFrom(Filesystem::fileReadToEnd(path), pkey_type_);
-    loaded_ = true;
-    return;
-  }
+  issuer->pubkey_value = json_pubkey->getString("value", "");
+
   // Check "uri" and "cluster"
-  std::string uri = json_pubkey->getString("uri", "");
-  std::string cluster = json_pubkey->getString("cluster", "");
-  if (uri != "" && cluster != "") {
-    // Public key will be loaded from the specified URI.
-    uri_ = uri;
-    cluster_ = cluster;
-    return;
-  }
+  issuer->uri = json_pubkey->getString("uri", "");
+  issuer->cluster = json_pubkey->getString("cluster", "");
 
-  // Public key not found
-  ENVOY_LOG(debug, "IssuerInfo [name = {}]: Public key source missing", name_);
-  failed_ = true;
+  // Check "cache_expiration_sec".
+  issuer->pubkey_cache_expiration_sec = json_pubkey->getInteger(
+      "cache_expiration_sec", kPubKeyCacheExpirationSec);
 }
 
-bool IssuerInfo::IsAudienceAllowed(const std::string &aud) {
-  return audiences_.empty() || (std::find(audiences_.begin(), audiences_.end(),
-                                          aud) != audiences_.end());
+}  // namespace
+
+std::string IssuerConfig::Validate() const {
```

Comment:
Why don't you have this function return bool, and call ENVOY_LOG inside function to log error.

Reply:
I would say just throw exception in failure cases and make the function void.